File: backend/app/models/brainstorm.py
"""
Brainstorming models for AI-powered idea generation
Supports character, plot, world-building, and conflict brainstorming
"""

from sqlalchemy import Column, String, Text, DateTime, Integer, Float, Boolean, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid

from app.database import Base


class BrainstormSession(Base):
    """Brainstorming session tracker"""
    __tablename__ = "brainstorm_sessions"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    manuscript_id = Column(String, ForeignKey("manuscripts.id"), nullable=False)
    outline_id = Column(String, ForeignKey("outlines.id"), nullable=True)

    # Session metadata
    session_type = Column(String, nullable=False)  # 'CHARACTER', 'PLOT_BEAT', 'WORLD', 'CONFLICT'
    context_data = Column(JSON, default=dict)  # Genre, structure, existing entities, etc.

    # Status tracking
    status = Column(String, default='IN_PROGRESS')  # IN_PROGRESS, COMPLETED, ABANDONED

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    manuscript = relationship("Manuscript", backref="brainstorm_sessions")
    outline = relationship("Outline", backref="brainstorm_sessions")
    ideas = relationship("BrainstormIdea", back_populates="session", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<BrainstormSession(id={self.id}, type='{self.session_type}', status='{self.status}')>"


class BrainstormIdea(Base):
    """Individual ideas generated during brainstorming"""
    __tablename__ = "brainstorm_ideas"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id = Column(String, ForeignKey("brainstorm_sessions.id"), nullable=False)

    # Idea content
    idea_type = Column(String, nullable=False)  # Matches session_type
    title = Column(String, nullable=False)  # Short title/name
    description = Column(Text, nullable=False)  # Full AI-generated content
    idea_metadata = Column(JSON, default=dict)  # Type-specific attributes (want, need, flaw, etc.)

    # User actions
    is_selected = Column(Boolean, default=False)
    user_notes = Column(Text, default="")
    edited_content = Column(Text, nullable=True)  # User's edited version

    # Integration tracking
    integrated_to_outline = Column(Boolean, default=False)
    integrated_to_codex = Column(Boolean, default=False)
    plot_beat_id = Column(String, ForeignKey("plot_beats.id"), nullable=True)
    entity_id = Column(String, ForeignKey("entities.id"), nullable=True)

    # AI metadata
    ai_cost = Column(Float, default=0.0)
    ai_tokens = Column(Integer, default=0)
    ai_model = Column(String, default="anthropic/claude-3.5-sonnet")

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    session = relationship("BrainstormSession", back_populates="ideas")
    plot_beat = relationship("PlotBeat", foreign_keys=[plot_beat_id])
    entity = relationship("Entity", foreign_keys=[entity_id])

    def __repr__(self):
        status_icon = "✓" if self.is_selected else "○"
        return f"<BrainstormIdea(id={self.id}, title='{self.title}', {status_icon})>"

File: frontend/src/types/brainstorm.ts
/**
 * Brainstorming & Ideation Tool Types
 * Matches backend models from app/models/brainstorm.py
 */

export type BrainstormSessionType = 'CHARACTER' | 'PLOT_BEAT' | 'WORLD' | 'CONFLICT' | 'SCENE';
export type BrainstormSessionStatus = 'IN_PROGRESS' | 'COMPLETED' | 'ABANDONED';
export type BrainstormIdeaType = 'CHARACTER' | 'PLOT_BEAT' | 'WORLD' | 'CONFLICT' | 'SCENE' | 'CHARACTER_WORKSHEET';

export interface BrainstormSession {
  id: string;
  manuscript_id: string;
  outline_id?: string;
  session_type: BrainstormSessionType;
  context_data: Record<string, any>;
  status: BrainstormSessionStatus;
  created_at: string;
  updated_at: string;
}

export interface BrainstormIdea {
  id: string;
  session_id: string;
  idea_type: string;
  title: string;
  description: string;
  idea_metadata: Record<string, any>;
  is_selected: boolean;
  user_notes: string;
  edited_content?: string;
  integrated_to_outline: boolean;
  integrated_to_codex: boolean;
  plot_beat_id?: string;
  entity_id?: string;
  ai_cost: number;
  ai_tokens: number;
  ai_model: string;
  created_at: string;
  updated_at: string;
}

// Character-specific metadata (Brandon Sanderson methodology)
export interface CharacterMetadata {
  name: string;
  role: string;
  want: string;
  need: string;
  flaw: string;
  strength: string;
  arc: string;
  hook: string;
  relationships: string;
  story_potential: string;
}

// Session statistics
export interface BrainstormSessionStats {
  session_id: string;
  session_type: BrainstormSessionType;
  status: BrainstormSessionStatus;
  total_ideas: number;
  selected_ideas: number;
  integrated_ideas: number;
  total_cost: number;
  total_tokens: number;
  created_at: string;
}

// API request/response types
export interface CreateSessionRequest {
  manuscript_id: string;
  outline_id?: string;
  session_type: BrainstormSessionType;
  context_data: Record<string, any>;
}

export interface CharacterGenerationRequest {
  api_key: string;
  genre: string;
  story_premise: string;
  character_ideas?: string;
  num_ideas?: number;
}

export interface PlotGenerationRequest {
  api_key: string;
  genre: string;
  premise: string;
  num_ideas?: number;
}

export interface LocationGenerationRequest {
  api_key: string;
  genre: string;
  premise: string;
  num_ideas?: number;
}

export interface UpdateIdeaRequest {
  is_selected?: boolean;
  user_notes?: string;
  edited_content?: string;
}

export interface IntegrateCodexRequest {
  entity_type?: string;
}

// New Phase 5 request types
export interface RefineIdeaRequest {
  api_key: string;
  feedback: string;
  direction: 'refine' | 'expand' | 'contrast' | 'combine';
  combine_with_idea_id?: string;
}

export interface ConflictGenerationRequest {
  api_key: string;
  genre: string;
  premise: string;
  character_ids?: string[];
  conflict_type: 'internal' | 'interpersonal' | 'external' | 'societal' | 'any';
  num_ideas?: number;
}

export interface SceneGenerationRequest {
  api_key: string;
  genre: string;
  premise: string;
  beat_id?: string;
  character_ids?: string[];
  location_id?: string;
  scene_purpose: 'introduction' | 'conflict' | 'revelation' | 'climax' | 'resolution' | 'any';
  num_ideas?: number;
}

export interface CharacterWorksheetRequest {
  api_key: string;
  character_id?: string;
  character_name?: string;
  worksheet_type: 'full' | 'backstory' | 'psychology' | 'voice' | 'relationships';
}

export interface ExpandEntityRequest {
  api_key: string;
  expansion_type: 'deepen' | 'relationships' | 'history' | 'secrets' | 'conflicts';
}

// Conflict metadata
export interface ConflictMetadata {
  title: string;
  type: 'internal' | 'interpersonal' | 'external' | 'societal';
  participants: string[];
  core_tension: string;
  stakes: string;
  trigger: string;
  escalation_points: string[];
  possible_resolutions: string[];
  emotional_core: string;
  scene_opportunities: string[];
  thematic_connection: string;
}

// Scene metadata
export interface SceneMetadata {
  title: string;
  purpose: 'introduction' | 'conflict' | 'revelation' | 'climax' | 'resolution';
  pov_character: string;
  characters_present: string[];
  location: string;
  opening_hook: string;
  scene_goal: string;
  obstacle: string;
  outcome: 'disaster' | 'partial_success' | 'success' | 'twist';
  emotional_arc: {
    start: string;
    shift: string;
    end: string;
  };
  scene_beats: string[];
  sensory_details: string[];
  dialogue_moments: string[];
  subtext: string;
}

// Character worksheet response
export interface CharacterWorksheet {
  character_name: string;
  worksheet_type: string;
  sections: Record<string, any>;
  _metadata: {
    character_id?: string;
    idea_id?: string;
    ai_cost: number;
    ai_tokens: number;
  };
}

// Entity expansion response
export interface EntityExpansion {
  expanded_attributes: Record<string, any>;
  suggestions: string[];
  story_hooks: string[];
  _expansion_metadata: {
    type: string;
    ai_cost: number;
    ai_tokens: number;
  };
}

// Related entity response
export interface RelatedEntity {
  type: string;
  name: string;
  relationship_to_source: string;
  hook: string;
  [key: string]: any;
  _metadata: {
    source_entity_id: string;
    relationship_type: string;
    ai_cost: number;
  };
}

// UI-specific types
export type BrainstormTechnique = 'character' | 'plot' | 'world' | 'conflict';

export interface BrainstormModalState {
  isOpen: boolean;
  technique?: BrainstormTechnique;
  manuscriptId?: string;
  outlineId?: string;
  plotBeatId?: string;
}

// Manuscript context for auto-loading
export interface BrainstormContext {
  outline: {
    genre: string | null;
    premise: string | null;
    logline: string | null;
  };
  existing_entities: {
    characters: { id: string; name: string }[];
    locations: { id: string; name: string }[];
  };
}

// Session with preview and stats
export interface SessionWithPreview {
  session: BrainstormSession;
  preview_ideas: {
    id: string;
    title: string;
    description: string;
  }[];
  stats: {
    total_ideas: number;
    integrated_count: number;
    pending_count: number;
  };
}
